framework/rendering/hpp_render_context.cpp

```diff
@@ -302,16 +302,24 @@ void HPPRenderContext::begin_frame()
 
 		if (result == vk::Result::eSuboptimalKHR || result == vk::Result::eErrorOutOfDateKHR)
 		{
+#if defined(PLATFORM__MACOS)
+			// On Apple platforms, force swapchain update on both eSuboptimalKHR and eErrorOutOfDateKHR
+			// eSuboptimalKHR may occur on macOS/iOS following changes to swapchain other than extent/size
+			bool swapchain_updated = handle_surface_changes(true);
+#else
 			bool swapchain_updated = handle_surface_changes(result == vk::Result::eErrorOutOfDateKHR);
+#endif
 
 			if (swapchain_updated)
 			{
+				// Need to destroy and reallocate acquired_semaphore since it may have already been signaled
+				device.get_handle().destroySemaphore(acquired_semaphore);
```

Comment:
Don't know exactly under what circumstances this code path is taken.
But simply destroying the semaphore that you've got via `prev_frame.request_semaphore_with_ownership()` seems odd. Maybe you should call `prev_frame.release_owned_semaphore()` here?

Reply:
> Don't know exactly under what circumstances this code path is taken

This is existing code that is run on any platform where `VK_ERROR_OUT_OF_DATE_KHR` is returned by `swapchain->acquire_next_image()`.  I am simply extending that case on Apple platforms to also cover when `VK_SUBOPTIMAL_KHR` is returned.

> Maybe you should call prev_frame.release_owned_semaphore() here?

Funny thing, I initially tried exactly that until I realized that all it does is return the signaled semaphore to the `released_semaphores` vector in `SemaphorePool` - it does not unsignal  the semaphore.  When that signaled semaphore is picked up again it will cause the same problem inside `[HPP]RenderContext::begin_frame()`.  I am not aware of any mechanism to manually clear a Vulkan binary semaphore (other than wait on a queue), so my only option was to destroy it and get another.  Fortunately the semantics of `request_semaphore_with_ownership()` means that it is manually-managed, and not present in any private vector within `SemaphorePool` until such time as you return it to the pool via `release_owned_semaphore()` in `[HPP]RenderContext::end_frame()`.  If you know of a mechanism to manually unsignal a binary semaphore please let me know.

FYI, here is the Validation error message if you don't use an unsignaled semaphore:

`[error] -370888023 - VUID-vkAcquireNextImageKHR-semaphore-01286: Validation Error: [ VUID-vkAcquireNextImageKHR-semaphore-01286 ] Object 0: handle = 0xdf43f6000000012b, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0xe9e4b2a9 | vkAcquireNextImageKHR():  Semaphore must not be currently signaled. The Vulkan spec states: If semaphore is not VK_NULL_HANDLE it must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.290.0/mac/1.3-extensions/vkspec.html#VUID-vkAcquireNextImageKHR-semaphore-01286)`

Reply:
@asuessenbach I believe your point is resolved with explanation above.  Please let me know if otherwise.  Thx.

Reply:
Ah, yes, you're right. Thanks a lot for your explanations.

That is, from Windows-point-of-view, it looks good to me. Can't tell anything about the Mac-issue, though.